src/tests/suite/daos_nvme_recovery.c

```diff
@@ -119,9 +119,263 @@ nvme_recov_1(void **state)
+ * daos_mgmt_get_bs_state() C API.
+ */
+static void
+nvme_test_get_blobstore_state(void **state)
+{
+	test_arg_t	*arg = *state;
+	device_list	*devices = NULL;
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+	char		 dkey[DTS_KEY_LEN] = { 0 };
+	char		 akey[DTS_KEY_LEN] = { 0 };
+	int		 obj_class, key_nr = 10;
+	int		 rank = 0, tgt_idx = 0;
+	int		 blobstore_state;
+	int		 i, j;
+	int		 ndisks;
+	int		 retry_cnt;
+	int		 rc;
+
+	if (!is_nvme_enabled(arg)) {
+		print_message("NVMe isn't enabled.\n");
+		skip();
+	}
+
+	/**
+	 * Get the total number of NVMe devices from all the servers.
+	 */
+	rc = dmg_storage_device_list(dmg_config_file, &ndisks, NULL);
+	assert_int_equal(rc, 0);
+	print_message("Total Disks = %d\n", ndisks);
+
+	/**
+	 * Get the Device info of all NVMe devices.
+	 */
+	D_ALLOC_ARRAY(devices, ndisks);
+	rc = dmg_storage_device_list(dmg_config_file, NULL, devices);
+	assert_int_equal(rc, 0);
+	for (i = 0; i < ndisks; i++)
+		print_message("Rank=%d UUID=%s state=%s host=%s\n",
+			      devices[i].rank, DP_UUID(devices[i].device_id),
+			devices[i].state, devices[i].host);
+
+
+	/**
+	 * Set the object class and generate data on objects.
+	 */
+	if (arg->pool.pool_info.pi_nnodes < 2)
+		obj_class = DAOS_OC_R1S_SPEC_RANK;
+	else
+		obj_class = DAOS_OC_R2S_SPEC_RANK;
+
+	oid = dts_oid_gen(obj_class, 0, arg->myrank);
+	oid = dts_oid_set_rank(oid, rank);
+	oid = dts_oid_set_tgt(oid, tgt_idx);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+
+	print_message("Generating data on obj "DF_OID"...\n", DP_OID(oid));
+	for (i = 0; i < key_nr; i++) {
+		dts_key_gen(dkey, DTS_KEY_LEN, "dkey");
+		for (j = 0; j < key_nr; j++) {
+			dts_key_gen(akey, DTS_KEY_LEN, "akey");
+			insert_single(dkey, akey, 0, "data", strlen("data") + 1,
+				      DAOS_TX_NONE, &req);
+		}
+	}
+	ioreq_fini(&req);
+
+	/**
+	 * Verify blobstore of first device returned is in "NORMAL" state
+	 * before setting to faulty.
+	 */
+	rc = daos_mgmt_get_bs_state(arg->group, devices[0].device_id,
+				    &blobstore_state, NULL /*ev*/);
+	assert_int_equal(rc, 0);
+
+	rc = verify_blobstore_state(blobstore_state, "normal");
+	assert_int_equal(rc, 0);
+	print_message("Blobstore is in NORMAL state\n");
+
+	/**
+	 * Manually set first device returned to faulty via
+	 * 'dmg storage set nvme-faulty'.
+	 */
+	print_message("NVMe with UUID=%s on host=%s\" set to Faulty\n",
+			DP_UUID(devices[0].device_id),
+			devices[0].host);
+	rc = dmg_storage_set_nvme_fault(dmg_config_file, devices[0].host,
+					devices[0].device_id, 1);
+	assert_int_equal(rc, 0);
+
+	/**
+	 *  Continue to check blobstore state until "OUT" state is returned
+	 *  or max test retry count is hit (5 min).
+	 */
+	retry_cnt = 0;
+	while (retry_cnt <= 15) {
+		rc = daos_mgmt_get_bs_state(arg->group, devices[0].device_id,
+					    &blobstore_state, NULL /*ev*/);
+		assert_int_equal(rc, 0);
+
+		if (verify_blobstore_state(blobstore_state, "out") == 0)
+			break;
+
+		sleep(20);
+		retry_cnt++;
+	};
```

Comment:
Suggestion: I suppose "verify bs state" will be used widely in many test cases, we may introduce a function like "int wait_bs_state(rank, tgt_idx, expected state, timeout)".